Client performance solr reindex

ktisdell opened the pull request:
**Link to QA issue**
https://github.com/BroadleafCommerce/QA/issues/3922

**Link to FTD-specific installation instructions**
https://blc.slack.com/files/T02654EKZ/FS4C64H8B

ktisdell committed:
This commit is to add new components to better coordinate the indexing
and reindexing of Solr and to make things safer and more efficient.
These are new components to avoid changing what is already available.

ktisdell committed:
Adding some retry logic for Zookeeper interactions.

ktisdell committed:
Making the Hibernate utils more generic with respect to persistence
operations.

ktisdell committed:
Making use of a more generic operation construct.

ktisdell committed:
Added some documentation around the ZK distributed lock.

ktisdell committed:
Fixing / updating retryable interactions.

ktisdell committed:
Simplifying the exception details.

ktisdell committed:
Simplified the exception handling in the API.

ktisdell committed:
Renaming the operation util.

ktisdell committed:
Fixed issue with lock monitor where we were using the wrong monitor.

ktisdell committed:
Added additional docs.

ktisdell committed:
Updates to the Zookeeper lock.

ktisdell committed:
Refactoring slightly.

ktisdell committed:
Added a blocking queue, backed by Zookeeper.

ktisdell committed:
Added some documentation and made some minor adjustments.

ktisdell committed:
Added additional documenation.

ktisdell committed:
Fixed bugs and performance.

ktisdell committed:
More documentation, bug fixes, performance improvements.

ktisdell committed:
Added additional documentation.

ktisdell committed:
Setting the default max queue size to 500.

ktisdell committed:
Added additional metadata and convenience methods.

ktisdell committed:
Updated Java docs.

ktisdell committed:
Refactoring and implementing changes to allow for full reindexing of the
catalog.

ktisdell committed:
Added a mechanism to run with an entity manager in a background thread.
Started implementing the CatalogSolrIndexUpdateCommandHandler. Added a
cross-thread state management mechanism.

ktisdell committed:
Continuing to develop a multi-threaded Solr reindexing solution.

ktisdell committed:
Refactored Zookeeper Queue and Lock to be more generic and depend less
on Solr-specific Zookeeper APIs.

ktisdell committed:
Continuing to refactor and add support for faster, more flexible, and
more controlled catalog indexing in Solr.

ktisdell committed:
Added extension hooks for pre/post processing of events.

ktisdell committed:
Added license headers to new classes.

ktisdell committed:
Deleted unnecessary classes.  Added some new convenience methods to
HibernateUtils.

ktisdell committed:
Added a method to SolrIndexService to indicate if legacy mode should be
used.  Added default property to common.properties for legacy mode.
Refactored DefaultQueueProviderImpl to allow optional constructor args.
Added default commands to CatalogSolrIndexUpdateService for catalog and
site reindexing.

ktisdell committed:
Refactoring the command handler a little bit to avoid recursive calls.
Adding some additional default configs.

ktisdell committed:
Better error handling.  Added a marker interface.

ktisdell committed:
Added license header.

ktisdell committed:
Refactored APIs.

ktisdell committed:
Incremental updates, refacotring, and bug fixes.

ktisdell committed:
The BroadleafRequestContext automatically creates a BLC and binds it to
a thread, even if one is not required.  It makes it difficult for a
process, especially a background process to know if it should remove
that BLC.  This commit doesn't change default behavior, but makes it
more clear and flexible.

ktisdell committed:
Made the BroadleafRequestContext a little easier to use in background
threads.  The ThreadLocal was set up to create a BroadleafRequestContext
upon request, so you never knew if you just created it or if it was
there prior.  This was OK when you have a filter chain that handles
this, but when running in background threads, it's hard to know if /
when a BRC was created and by whom.  This should fix it.

ktisdell committed:
Incremental update / bug fixes.

ktisdell committed:
Refactored to avoid duplicate logic.  Fixed bugs / issues. Added
comments.

ktisdell committed:
Added an isSuccessful method to allow a custom hook to determine if the
process was successful prior to swapping cores, for example.

ktisdell committed:
Refactored the building of Solr documents into its own component. Added
Javadocs.

ktisdell committed:
This commit attempts to bridge the gap between the existing Solr search
implementation(s) and the new Solr indexer implementation.  The new Solr
indexer makes use of SolrJ's (newer) optional ability to specify the
collection / core on each method invocation that interacts with Solr.
This was available on CloudSolrClient, but it is available on
HttpSolrClient as well.  As a result, the new indexer explicitly
provides the foreground and background collection names when interacting
with the SolrJ client.  This causes a problem with the legacy search
code that does not make use of these explicit collection names,
especially the HttpSolrClient, which specifies a base URL. For example,
if the base URL is http://localhost:8983/solr, then it works great for
the new indexer, which passes in a collection such as "catalog_reindex",
and the client interacts with
http://localhost:8983/solr/catalog_reindex.  However, it does not work
well for the legacy search functionality which would interact with
http://localhost:8983/solr which is not a valid collection URL.
Optionally, if the base URL is
http://localhost:8983/solr/catalog_reindex, then the new indexer will
pass in "catalog_reindex" and the client will try to interact with
http://localhost:8983/solr/catalog_reindex/catalog_reindex, which is
also not valid.  These changes attempt to bridge this gap.

ktisdell committed:
Decoupling the DelegatingHttpSolrClient from SolrConfiguration a bit.
Also, making it so that the default is that a command lock cannot be
obtained by default, so that one has to explicitly enable that in an
environment

ktisdell committed:
Fixed issue where we weren't checking the method to see if things are
successful.

ktisdell committed:
Fixed a small bug related to wait times.  Added some documentation.
Cleaned a few things up.

ktisdell committed:
Added a default property for globally obtaining a distributed lock.

ckittrell committed:
Adding logging for the case that a command thread is already registered for the provided command group & add additional comments describing the timing of when incremental commands are allowed relative to a FullReindexCommand

ckittrell commented on core/broadleaf-framework/src/main/java/org/broadleafcommerce/core/search/service/solr/indexer/AbstractSolrIndexUpdateServiceImpl.java:
```diff
@@ -0,0 +1,292 @@
+                    // be overridden when the full reindex is applied. Just before the FullReindexCommand begins, it
+                    // is removed from the queue.  This allows incremental update commands to be queued with the assumption
+                    // that the full reindex may be mid-stream & may not have picked up the update.
+                    if (!commandQueue.contains(FullReindexCommand.DEFAULT_INSTANCE) && !commandQueue.contains(command)) {
```
@ktisdell I'm beginning to question this condition - do we really want to avoid queueing incremental update commands if a FullReindexCommand is in queue?

I think that logic is totally fine if the full reindex is successful, but if the full reindex fails, then you're left with a foreground collection that won't be updated.

On the other hand, if we _do_ queue the incremental updates anyway, then they'll repeat a bit of the same work, but it _should be_ harmless.

ktisdell replied:
@ckittrell I don't disagree and don't have a strong opinion here.  I think more work could be done to have a coordinated foreground command queue and background command queue which could effectively handle situations like that.  Given that we don't, I'm OK with removing that logic in favor of "the extra work" that will have to be done by the indexer, assuming that doing that extra work simply restores the state after it is complete (i.e. causes no harm).

ckittrell replied:
Agreed - lets hit the easy button here & remove the `!commandQueue.contains(FullReindexCommand.DEFAULT_INSTANCE)` portion of the condition.

ckittrell commented on core/broadleaf-framework/src/main/java/org/broadleafcommerce/core/search/service/solr/indexer/AbstractSolrIndexUpdateServiceImpl.java:
```diff
@@ -0,0 +1,292 @@
+    @Override
+    public SolrInputDocument buildDocument(Indexable indexable, List<IndexField> fields, List<Locale> locales) {
+        return commandHandler.buildDocument(indexable, fields, locales);
+    }
```
It seems like these `buildDocument(...)` methods don't belong in this class - I'd expect to only see them in the CommandHandler. @ktisdell do you have a particular use case in mind for these methods?

ktisdell replied:
@ckittrell I originally added them so that services outside of this class can build documents as needed.  Consider, for example, `WorkflowSolrIndexServiceImpl#getSolrDocuments(...)`.  In fact, this should likely be updated to optionally use this class for building documents rather than using the legacy class.

ktisdell replied:
Although, they could likely just use the `CatalogDocumentBuilder` as well.

ktisdell committed:
Added some validations to ensure that SKU browsing and Site Collections
are not expected when using the new Solr Indexer.

ktisdell committed:
Added facilities to allow external callers to perform CacheOperations,
which are internal catalog cache operations (different than Hibernate
cache, etc.)

ckittrell commented on core/broadleaf-framework/src/main/java/org/broadleafcommerce/core/search/service/solr/indexer/AbstractSolrIndexUpdateCommandHandlerImpl.java:
```diff
@@ -0,0 +1,271 @@
+            
+            try {
+                if (changeMade) {
+                    commit(collectionName, true, true, false);
```
While I agree that committing all deletes & additions at once is the right idea, there is potential for a huge commit if the command holds a large number of deletions and/or additions. 

Splitting into sets of smaller commits is likely the wrong idea since rollback scenarios get ugly quickly.

The best answer may be to enforce a limit on the number of deletions & additions per command. From there, it would be the responsibility of the caller to provide smaller commands.

Of course, this is assuming that there's a relatively low threshold for the max number of items that can be committed before seeing a significant performance degradation.

ckittrell committed:
Remove the !commandQueue.contains(FullReindexCommand.DEFAULT_INSTANCE) which would cause foreground collection updates to be ignored if the full reindex failed

Commit:
Fix SolrConfiguration to set Primary Name when solrServer is DelegatingHttpSolrClient

Commit:
Actually fix collection names

ckittrell committed:
Readd line to ensure that ignoreSite is correctly declared on the BRC
